src/parser/parser_flow.ml

```diff
@@ -2556,7 +2565,18 @@ end = struct
               body;
             }))
         | _ ->
+            raise Try.Rollback
+      in
+
+      fun env ->
+        let start_loc = Peek.loc env in
+        Expect.token env T_FOR;
+        Expect.token env T_LPAREN;
+        match Try.to_parse env (try_forin_or_forof start_loc) with
```

Comment:
From the parser's point of view, it can use the existence of an uninitialized init as evidence that this MUST be a for-of or a for-in. So if you have a pattern with uninitialized init, and then come across a `;` (indicating a regular for loop) where you expect an `in` or `of`, the parser can throw an unexpected token.

If you want to give a better error, then yeah, you can do a little more work to keep track of where there is an uninitialized init.

Either way, you can probably just make `variable_declaration_list` return a list of the uninitialized  init errors and decide later whether to throw them.